### ISSUE TITLE
Refactor GameWindow out of PyBoy class and into MB

### ISSUE THREAD

**Issue by Baekalfen** [Baekalfen/PyBoy#65]

**Comment by Baekalfen** [comment]
`pyboy.core.lcd` is a class separate to `pyboy.window` becausa the LCD is in the "guest" part of the emulation and the window is in the "host" part. The LCD contains registers and memory associated with the graphics module of the Game Boy. But rendering that graphics on a modern computer, is completely foreign to the Game Boy. This is the way all modules in PyBoy are (should be) implemented.

The window also serves two other purposes than being the LCD class. It receives keyboard input for the `Interaction` class and synchronizes the emulated time to the real-world time through the frame-limiter.

If we moved the window rendering code to the LCD, we also would have to choose a single way to render the Game Boy's graphics onto a frame buffer. Currently we already have two different ways ― `scanline` and `SDL2` ― with different advantages.

Now that I'm thinking about it, I don't think I would want what the title says.

> so PyBoy could run silently in the background and still be fully functional for both input and output via the public API

We can do that today.

Or are you saying we should implement the "window" through the public API?

**Comment by krs013** [comment]
I agree that there is and should be a distinction between the host and guest parts of emulation, but I think that right now we have put parts of the guest into the host with the rendering. The host/ui should be responsible for receiving input from the keyboard (but not the API--we have it right currently with `interaction.py` living in `core`) and displaying the screen. However, the guest/emulator/core should be responsible for determining what to display on the screen. 

Right now, the `window` classes have `scanline` and `render_screen` methods that do a significant amount of emulation that pertains to how the gameboy works. From a design standpoint, I think that process should be in `core` and not part of the UI. Whether we have multiple implementations of those processes is not the issue. We could just as easily have two implementations of `lcd.py` (or `core/render.py` if we want to keep the memory separate from the functionality but I don't think that helps us). 

Making this change would be a technical challenge, because both version of our render pipelines are at least a little integrated with SDL2, though not as much as I thought. For both of them, there is a copy of the screen buffer in memory owned by SDL2, and our code updates the screen buffer before blitting the screen. `window_sdl2.py` has its own copy of the screen buffer in memory, which it updates and then copies to SDL2's copy once per frame. This would be very easy to separate. `window_scanline.py` does not keep a copy of the screen buffer, and only copies one line at a time to SDL2's screen buffer. This is the reason that `window_scanline` cannot implement any of the `get_screen_buffer` functionality. Since that doesn't seem to give it much or any of a performance advantage over the other method, I think it would be fine to change that.

So what I'm proposing is that the rendering code and python-space screen buffers be moved out of the windows and interact with any SDL2 code through a `ui` module or the like. The `ui` module can provide a method hook that is called with a reference to the screen buffer to copy data over to SDL2 each frame. This hook could also then be populated with user functions instead, making the public API more usable for custom applications. So in a sense, this is like implementing the window through the public API, though there is a distinction between having the `lcd` code calling render on the window, and the window calling `get_screen_buffer` on the `lcd` module.  But really, I think either of those options would be fine.

Finally, with respect to PyBoy "running silently in the background" and all... yes, we can do that right now, but it involves opening an SDL2 window and then hiding it. The SDL2 side of it doesn't actually do any meaningful work in running PyBoy invisibly, though, and may even reduce performance. Under my model, the SDL2 window would never be instantiated, and every frame rendered by `lcd` would only go to the public API or screen recorder or whatever we want with it.

PS--Also, I don't think it really helps us to have two different render pipelines. `window_sdl2.py` has the advantage that it has its own screen buffer and is sometimes faster, while `window_scanline.py` produces more correct output and usually a more stable framerate because it doesn't have to rebuild the cache when tiles are changed. I've always wanted to eventually merge them, but that doesn't have to happen now. If we want, though, we could take this as an opportunity to merge the methods, using the scanline render code but adding caching and a local screen buffer. It's worth pointing out that this will still allow us to have the option of using SDL2 or OpenGL for the window part if we want to add shaders or the like.

**Comment by Baekalfen** [comment]
> we have put parts of the guest into the host with the rendering

Would you point out an example?

> However, the guest/emulator/core should be responsible for determining what to display on the screen

I mean, it does from the guest's point of view. How the `BaseWindow` subclasses chooses to implement it on the host side, is hard to let the guest part control in a meaningful way. And to your point, that's why I made the `LCDCRegister` and `PaletteRegisters`. So this would be determined by the LCD and not the window implementation.

> the window classes have scanline and render_screen methods that do a significant amount of emulation

The rendering is implemented with knowledge of the Game Boy, yes. But how would you abstract it away in a way that will work for the both of the `scanline` and `SDL2` implementations? About the `scanline` function, I would be fine with moving that to `lcd.py`, but it's only really used in my window implementations.

> So what I'm proposing is that the rendering code and python-space screen buffers be moved out of the windows and interact with any SDL2 code through a ui module or the like.

Wouldn't it be easier to make a class hierarchy of `BaseWindow`->`SDLBackend`->`SDL2|scanline` and just move the common functionality in the `SDLBackend`? If there is any of the functionality in the windows that belongs in `lcd.py`, we could move them at the same time.

> it involves opening an SDL2 window and then hiding it. The SDL2 side of it doesn't actually do any meaningful work in running PyBoy invisibly

I'm not sure what you mean. The point is just to run the tests without the windows keep popping up. It's not meant to be used for much else. If SDL2 doesn't really do anything, that's to a certain degree fine. We are mostly interested in testing the code in our codebase.

> may even reduce performance

Compared to showing the window?

> Under my model, the SDL2 window would never be instantiated, and every frame rendered by lcd would only go to the public API or screen recorder or whatever we want with it.

That's more or less the use of the `headless` window. All of the tests use `headless` because it fully works without displaying anything on the host. It's only the actual tests of the different windows that use the `scanline`, `OpenGL`, `SDL2`, `headless`, and `dummy` classes.

>  I don't think it really helps us to have two different render pipelines

I'm a bit indifferent. You're welcome to merge them if you want. It's just important, that if we use the `scanline` code, that we will still be able to get the screen buffer, as `ScreenRecorder` and the tests depend on it. And then of course not negatively affect performance, if possible.



So in conclusion, I think many of the things you ask for could be handled with minimal changes, if you find which features of the window implementations that we would want to move to `lcd.py`, and then we move the common SDL code into a `SDLBackend` parent class.

**Comment by krs013** [comment]
By "we have put parts of the guest into the host with the rendering" I meant the `scanline` method of `window_scanline.py` and the `render_screen` method of both windows.

It seems we disagree on this, but I'm convinced that turning the LCD memory into pixels is part of the emulation and belongs in `core`, just as CPU emulation and (eventually) turning sound registers into samples are part of the emulation and not part of the host program that contains them. The video hardware was an internal part of the SoC bundled with the CPU and other modules, and it doesn't make sense to me to treat that differently just because it's not visible to software running on it.

Maybe we should discuss this: is PyBoy a platform that emulates the operation of Game Boy *hardware*, or does it emulate the running of Game Boy *software*? It's been my impression that we were pursuing the former, but if it's the latter then that might change my point of view. I get that that's an odd distinction but I hope it makes sense.

Consider this: if we add Game Boy Color support in the future, is that a feature that changes the core of the emulator or just the environment it runs in? (Ignore for that the fact that there are some changes to CPU operation, I'm mainly interested in the video aspect). From my point of view, this is very much a changing of the core, and the environment that contains it should look mostly or entirely the same.

WRT the `PaletteRegister`... I actually think that belongs in the host/window/UI and not the guest/core. Also the fact that `mb.py` initializes it with a default value that comes from `base_window.py` but the value is still passed to `lcd.py` is too much coupling for me.

> But how would you abstract it away in a way that will work for the both of the scanline and SDL2 implementations?

It would be sufficient to just have a `render_screen` function or something that accepts a reference or pointer to an internal screen buffer owned by `lcd.LCD` and copies it to the screen. `window_SDL2` already has this interface, and it would be very simple to add this to `window_scanline` as well (and may even make it faster...). I couldn't get SDL2 palettes/indexed colors to work before, but it may be an option, so the screen buffer could be 4-bit or 8-bit color indices instead of 32-bit colors, and then the window can map those to the actual colors, allowing for green/sepia effects if desired.

> About the scanline function, I would be fine with moving that to lcd.py, but it's only really used in my window implementations.

`window_scanline.py` uses `scanline` too, and does 99% of its rendering inside of it, so I'm not sure I understand what you mean.

> > may even reduce performance

> Compared to showing the window?

Yes, in the specific use case of rendering the screen without displaying it. The current setup is fine for running tests, but there may be users who want to run PyBoy headless (possibly even without SDL2 installed if they're on a system with no desktop like a VPS, where `SDL_Init` may return an error) and still get the video output through different means. I think we had a user who was making a twitch bot--that's the use case I'm imagining. And running the SDL2 methods in the background may still take resources even if the window is hidden.

WRT the render pipelines... I still would like to merge them eventually, but yeah, only if the features all stay and performance doesn't suffer. And that's not particularly important or relevant right now. The issue is about moving the `windows` into `mb` and I wanted to clarify and share my opinion on that, hehe. Sorry about all the long rants though! 

I think, given this discussion, that it's at least worth me taking a shot at it, since it really wouldn't be much work. I can make a branch/PR, and then based on how it looks we can consider whether to keep it. It'll probably also help make it more clear what I'm trying to describe, and highlight the advantages better.

**Comment by Baekalfen** [comment]
Introduced in `plugin_refactor` branch. It will be merged soon.
